les/api.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package les
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+var (
+	errNoStableCheckpoint = errors.New("no stable checkpoint provided")
+)
+
+// PublicLesServerAPI  provides an API to access the les server.
+// It offers only methods that operate on public data that is freely available to anyone.
+type PublicLesServerAPI struct {
+	server *LesServer
+}
+
+// NewPublicLesServerAPI creates a new les server API.
+func NewPublicLesServerAPI(server *LesServer) *PublicLesServerAPI {
```

Comment:
The API part of this PR is fine but please move it to a subsequent PR so that we can merge this one soon. This functionality depends on trusted servers sending new checkpoints regularly and it also makes sense only if the syncing part is implemented too. Implementing/merging in multiple steps is definitely desirable but we should not expose it through the API until it is actually functional.

Reply:
Will move these logics to a subsequent PR

Reply:
And maybe this part needs more detail annotations, if we can keep the logic here, i think more clear and detail annotations are needed.

Reply:
Since it is an admin function, shoudn't it be a private API instead of public? A private admin API might be ok to add at this point for convenient development reasons, even if the format might change in the final version.

Reply:
Yes, will make it a private API soon.